deepmd/pt/model/descriptor/env_mat.py

```diff
@@ -21,12 +21,15 @@ def _make_env_mat(
     nall = coord.shape[1]
     mask = nlist >= 0
     # nlist = nlist * mask  ## this impl will contribute nans in Hessian calculation.
-    nlist = torch.where(mask, nlist, nall - 1)
+    nlist = torch.where(mask, nlist, nall)
     coord_l = coord[:, :natoms].view(bsz, -1, 1, 3)
     index = nlist.view(bsz, -1).unsqueeze(-1).expand(-1, -1, 3)
-    coord_r = torch.gather(coord, 1, index)
+    coord_pad = torch.concat([coord, coord[:, -1:, :] + rcut], dim=1)
+    coord_r = torch.gather(coord_pad, 1, index)
     coord_r = coord_r.view(bsz, natoms, nnei, 3)
     diff = coord_r - coord_l
+    # avoid the possibility that coord[:, -1:, :] + rcut is the same as the coordinate of a real atom
+    diff = torch.where(torch.abs(diff) < 1e-30, torch.full_like(diff, 1e-30), diff)
     length = torch.linalg.norm(diff, dim=-1, keepdim=True)
```

Comment:
No need to hack the returned `diff`.

```suggestion
    diff_ = torch.where(torch.abs(diff) < 1e-30, torch.full_like(diff, 1e-30), diff)
    length = torch.linalg.norm(diff_, dim=-1, keepdim=True)
```